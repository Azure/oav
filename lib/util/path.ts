// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { flatMap } from "@ts-common/iterator"

export const splitPathAndReverse = (p: string | undefined) =>
  p === undefined ? undefined : Array.from(flatMap(p.split("/"), s => s.split("\\"))).reverse()

export const isSubPath = (
  mainPath: ReadonlyArray<string> | undefined,
  subPath: ReadonlyArray<string> | undefined,
) =>
  // return `true` if there are no subPath.
  subPath === undefined ||
  subPath.length === 0 ||
  // return `true` if `subPath` is a sub-path of `mainPath`.
  (mainPath !== undefined &&
    mainPath.length > subPath.length &&
    subPath.every((s, i) => mainPath[i] === s)
  )
