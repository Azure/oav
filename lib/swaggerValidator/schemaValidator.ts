import { JsonLoader } from "../swagger/jsonLoader";
import { Schema } from "../swagger/swaggerTypes";
import { strTemplate } from "../util/strTemplate";
import { ExtendedErrorCode, SourceLocation } from "../util/validationError";

export interface SchemaValidateContext {
  isResponse?: boolean;
  includeErrors?: ExtendedErrorCode[];
  statusCode?: string;
  jsonLoader?: JsonLoader;
}

export interface SchemaValidateIssue {
  code: ExtendedErrorCode;
  readonly message: string;
  readonly jsonPathsInPayload: string[];
  readonly schemaPath: string;
  readonly source: SourceLocation;
  readonly params?: any;
}

export type SchemaValidateFunction = (
  ctx: SchemaValidateContext,
  data: any
) => SchemaValidateIssue[];

export interface SchemaValidator {
  compile(schema: Schema): SchemaValidateFunction;
  compileAsync(schema: Schema): Promise<SchemaValidateFunction>;
  dispose(): Promise<void>;
}

export interface SchemaValidatorOption {
  isArmCall?: boolean;
}

const includeErrorsMap: WeakMap<ExtendedErrorCode[], Set<ExtendedErrorCode>> = new WeakMap();

export const getIncludeErrorsMap = (includeErrors?: ExtendedErrorCode[]) => {
  if (includeErrors === undefined) {
    return undefined;
  }
  let result = includeErrorsMap.get(includeErrors);
  if (result === undefined) {
    result = new Set(includeErrors);
    includeErrorsMap.set(includeErrors, result);
  }
  return result;
};

export const validateErrorMessages: { [key in ExtendedErrorCode]?: (params: any) => string } = {
  INVALID_RESPONSE_CODE: strTemplate`This operation does not have a defined '${"statusCode"}' response code`,
  INVALID_CONTENT_TYPE: strTemplate`Invalid Content-Type (${"contentType"}).  These are supported: ${"supported"}`,
  MISSING_REQUIRED_PARAMETER: strTemplate`Value is required but was not provided`,
  INVALID_RESPONSE_BODY: strTemplate`Body is required in response but not provided`,
  INVALID_RESPONSE_HEADER: strTemplate`Header ${"missingProperty"} is required in response but not provided`,
  MISSING_RESOURCE_ID: strTemplate`id is required to return in response of GET/PUT resource calls but not being provided`,
  LRO_RESPONSE_CODE: strTemplate`Respond to the initial request of a long running operation, Patch/Post call must return 201 or 202, Delete call must return 202 or 204, Put call must return 202 or 201 or 200, but ${"statusCode"} being returned`,
  LRO_RESPONSE_HEADER: strTemplate`Long running operation should return ${"header"} in header but not provided`,

  DISCRIMINATOR_VALUE_NOT_FOUND: strTemplate`Discriminator value "${"data"}" not found`,
  ANY_OF_MISSING: strTemplate`Data does not match any schemas from 'anyOf'`,
  ONE_OF_MISSING: strTemplate`Data does not match any schemas from 'oneOf'`,
  ONE_OF_MULTIPLE: strTemplate`Data is valid against more than one schema from 'oneOf'`,
  OBJECT_DEPENDENCY_KEY: strTemplate`Dependency failed - key must exist: ${"missingProperty"} (due to key: ${"property"})`,

  OBJECT_ADDITIONAL_PROPERTIES: strTemplate`Additional properties not allowed: ${"additionalProperty"}`,
  OBJECT_MISSING_REQUIRED_PROPERTY: strTemplate`Missing required property: ${"missingProperty"}`,
  OBJECT_PROPERTIES_MAXIMUM: strTemplate`Too many properties defined (${"data"}), maximum ${"limit"}`,
  OBJECT_PROPERTIES_MINIMUM: strTemplate`Too few properties defined (${"data"}), minimum ${"limit"}`,

  ARRAY_LENGTH_LONG: strTemplate`Array is too long (${"data"}), maximum ${"limit"}`,
  ARRAY_LENGTH_SHORT: strTemplate`Array is too short (${"data"}), minimum ${"limit"}`,
  ARRAY_UNIQUE: strTemplate`Array items are not unique (indexes ${"i"} and ${"j"})`,
  ARRAY_ADDITIONAL_ITEMS: strTemplate`Additional items not allowed`,

  INVALID_TYPE: strTemplate`Expected type ${"type"} but found type ${"data"}`,
  INVALID_FORMAT: strTemplate`Object didn't pass validation for format ${"format"}: ${"data"}`,
  PATTERN: strTemplate`String does not match pattern ${"pattern"}: ${"data"}`,
  MULTIPLE_OF: strTemplate`Value ${"data"} is not a multiple of ${"multipleOf"}`,
  ENUM_CASE_MISMATCH: strTemplate`Enum does not match case for: ${"data"}`,
  ENUM_MISMATCH: strTemplate`No enum match for: ${"data"}`,
  MAX_LENGTH: strTemplate`String is too long (${"data"} chars), maximum ${"limit"}`,
  MIN_LENGTH: strTemplate`String is too short (${"data"} chars), minimum ${"limit"}`,
  MINIMUM: strTemplate`Value ${"data"} is less than minimum ${"limit"}`,
  MAXIMUM: strTemplate`Value ${"data"} is greater than maximum ${"limit"}`,
  MINIMUM_EXCLUSIVE: strTemplate`Value ${"data"} is equal or less than exclusive minimum ${"limit"}`,
  MAXIMUM_EXCLUSIVE: strTemplate`Value ${"data"} is equal or greater than exclusive maximum ${"limit"}`,

  READONLY_PROPERTY_NOT_ALLOWED_IN_REQUEST: strTemplate`ReadOnly property "${"key"}" cannot be sent in the request`,
  WRITEONLY_PROPERTY_NOT_ALLOWED_IN_RESPONSE: strTemplate`Write-only property "${"key"}" is not allowed in the response`,
  SECRET_PROPERTY: strTemplate`Secret property "${"key"}" cannot be sent in the response`,
};

export const getValidateErrorMessage = (code: ExtendedErrorCode, param: any): string => {
  const func = (validateErrorMessages as any)[code];
  return func === undefined ? undefined : func(param);
};
