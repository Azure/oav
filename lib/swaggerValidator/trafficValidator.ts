import * as fs from "fs";
import * as path from "path";
import { resolve as pathResolve } from "path";
import { glob } from "glob";
import { toLower } from "lodash";
import {
  LiveValidationIssue,
  LiveValidator,
  RequestResponsePair,
} from "../liveValidation/liveValidator";
import { DefaultConfig } from "../util/constants";
import { apiValidationErrors, ErrorCodeConstants } from "../util/errorDefinitions";
import { OperationContext } from "../liveValidation/operationValidator";
import { traverseSwagger } from "../transform/traverseSwagger";
import { Operation, Path, LowerHttpMethods } from "../swagger/swaggerTypes";
import { LiveValidatorLoader } from "../liveValidation/liveValidatorLoader";
import { inversifyGetContainer, inversifyGetInstance } from "../inversifyUtils";

export interface TrafficValidationIssue {
  payloadFilePath?: string;
  specFilePath?: string;
  errors?: LiveValidationIssue[];
  operationInfo?: OperationContext;
  runtimeExceptions?: RuntimeException[];
}

export interface RuntimeException {
  code: string;
  message: string;
}

export interface OperationCoverageInfo {
  readonly spec: string;
  readonly coveredOperaions: number;
  readonly totalOperations: number;
  readonly coverageRate: number;
}

export class TrafficValidator {
  private liveValidator: LiveValidator;
  private trafficValidationResult: TrafficValidationIssue[] = [];
  private trafficFiles: string[] = [];
  private specPath: string;
  private trafficPath: string;
  private loader?: LiveValidatorLoader;
  private trafficOperation: Map<string, string[]> = new Map<string, string[]>();
  public operationSpecMapper: Map<string, string[]> = new Map<string, string[]>();
  public coverageResult: Map<string, number> = new Map<string, number>();
  public coverageData: OperationCoverageInfo[] = [];

  public constructor(specPath: string, trafficPath: string) {
    this.specPath = specPath;
    this.trafficPath = trafficPath;
  }

  public async initialize() {
    const specPathStats = fs.statSync(this.specPath);
    const trafficPathStats = fs.statSync(this.trafficPath);
    let specFileDirectory = "";
    let swaggerPathsPattern = "**/*.json";
    if (specPathStats.isFile()) {
      specFileDirectory = path.dirname(this.specPath);
      swaggerPathsPattern = path.basename(this.specPath);
    } else if (specPathStats.isDirectory()) {
      specFileDirectory = this.specPath;
    }
    if (trafficPathStats.isFile()) {
      this.trafficFiles.push(this.trafficPath);
    } else if (trafficPathStats.isDirectory()) {
      const searchPattern = path.join(this.trafficPath, "**/*.json");
      const matchedPaths = glob.sync(searchPattern, {
        nodir: true,
      });
      for (const filePath of matchedPaths) {
        this.trafficFiles.push(filePath);
      }
    }

    const liveValidationOptions = {
      checkUnderFileRoot: false,
      loadValidatorInBackground: false,
      directory: specFileDirectory,
      swaggerPathsPattern: [swaggerPathsPattern],
      excludedSwaggerPathsPattern: DefaultConfig.ExcludedExamplesAndCommonFiles,
      git: {
        shouldClone: false,
      },
    };

    this.liveValidator = new LiveValidator(liveValidationOptions);
    await this.liveValidator.initialize();

    const container = inversifyGetContainer();
    this.loader = inversifyGetInstance(LiveValidatorLoader, {
      container,
      fileRoot: liveValidationOptions.directory,
      ...liveValidationOptions,
      loadSuppression: Object.keys(apiValidationErrors),
    });

    const swaggerPaths = this.liveValidator.swaggerList;
    while (swaggerPaths.length > 0) {
      let swaggerPath = swaggerPaths.shift()!;
      let spec;
      try {
        spec = await this.loader.load(pathResolve(swaggerPath));
      } catch (e) {
        console.log(e);
      }
      if (spec !== undefined) {
        swaggerPath = toLower(swaggerPath);
        // Get Swagger - operation mapper.
        if (this.operationSpecMapper.get(swaggerPath) === undefined) {
          this.operationSpecMapper.set(swaggerPath, []);
        }
        traverseSwagger(spec, {
          onOperation: (operation: Operation, _path: Path, _method: LowerHttpMethods) => {
            if (
              operation.operationId !== undefined &&
              !this.operationSpecMapper.get(swaggerPath)?.includes(operation.operationId)
            ) {
              this.operationSpecMapper.get(swaggerPath)!.push(operation.operationId);
            }
          },
        });
      }
    }
  }

  public async validate(): Promise<TrafficValidationIssue[]> {
    let payloadFilePath;
    try {
      for (const trafficFile of this.trafficFiles) {
        payloadFilePath = trafficFile;
        const payload: RequestResponsePair = require(trafficFile);
        const validationResult = await this.liveValidator.validateLiveRequestResponse(payload);
        const operationInfo = validationResult.requestValidationResult?.operationInfo;
        const liveRequest = payload.liveRequest;
        const correlationId = liveRequest.headers?.["x-ms-correlation-request-id"] || "";
        const opInfo = await this.liveValidator.getOperationInfo(liveRequest, correlationId);
        let swaggerFile;
        if (liveRequest.url.includes("provider")) {
          swaggerFile = this.findSwaggerByOperationInfo(opInfo.info);
        } else {
          swaggerFile = this.findSwaggerByOperationId(opInfo.info);
        }
        if (swaggerFile !== undefined) {
          if (this.trafficOperation.get(swaggerFile) === undefined) {
            this.trafficOperation.set(swaggerFile, []);
          }
          if (!this.trafficOperation.get(swaggerFile)?.includes(opInfo.info.operationId)) {
            this.trafficOperation.get(swaggerFile)?.push(opInfo.info.operationId);
          }
        } else {
          console.log(`Error: Undefined operation ${JSON.stringify(opInfo.info)}`);
        }

        const errorResult: LiveValidationIssue[] = [];
        const runtimeExceptions: RuntimeException[] = [];
        if (validationResult.requestValidationResult.isSuccessful === undefined) {
          runtimeExceptions.push(validationResult.requestValidationResult.runtimeException!);
        } else if (validationResult.requestValidationResult.isSuccessful === false) {
          errorResult.push(...validationResult.requestValidationResult.errors);
        }

        if (validationResult.responseValidationResult.isSuccessful === undefined) {
          runtimeExceptions.push(validationResult.responseValidationResult.runtimeException!);
        } else if (validationResult.responseValidationResult.isSuccessful === false) {
          errorResult.push(...validationResult.responseValidationResult.errors);
        }
        if (errorResult.length > 0 || runtimeExceptions.length > 0) {
          this.trafficValidationResult.push({
            payloadFilePath,
            errors: errorResult,
            runtimeExceptions,
            operationInfo: operationInfo,
          });
        }
      }
    } catch (err) {
      const msg = `Detail error message:${err?.message}. ErrorStack:${err?.Stack}`;
      this.trafficValidationResult.push({
        payloadFilePath,
        runtimeExceptions: [
          {
            code: ErrorCodeConstants.RUNTIME_ERROR,
            message: msg,
          },
        ],
      });
    }
    this.operationSpecMapper.forEach((value: string[], key: string) => {
      if (this.trafficOperation.get(key) === undefined) {
        this.coverageResult.set(key, 0);
        this.coverageData.push({
          spec: key,
          coveredOperaions: 0,
          totalOperations: value.length,
          coverageRate: 0,
        });
      } else {
        if (value !== undefined && value.length !== 0) {
          this.coverageResult.set(key, this.trafficOperation.get(key)!.length / value.length);
          this.coverageData.push({
            spec: key,
            coveredOperaions: this.trafficOperation.get(key)!.length,
            totalOperations: value.length,
            coverageRate: this.trafficOperation.get(key)!.length / value.length,
          });
        } else {
          this.coverageResult.set(key, 0);
        }
      }
    });
    return this.trafficValidationResult;
  }

  private findSwaggerByOperationInfo(operationInfo: OperationContext) {
    let result = undefined;
    if (operationInfo.validationRequest === undefined) {
      return result;
    }
    for (const key of this.operationSpecMapper.keys()) {
      const value = this.operationSpecMapper.get(key);
      if (
        key.includes(toLower(operationInfo.apiVersion)) &&
        key.includes(toLower(operationInfo.validationRequest?.providerNamespace))
      ) {
        if (value!.includes(operationInfo.operationId)) {
          result = key;
          return result;
        }
      }
    }
    return result;
  }

  private findSwaggerByOperationId(operationInfo: OperationContext) {
    let result = undefined;
    for (const key of this.operationSpecMapper.keys()) {
      const value = this.operationSpecMapper.get(key);
      if (
        key.includes(toLower(operationInfo.apiVersion)) &&
        value!.includes(operationInfo.operationId)
      ) {
        result = key;
        return result;
      }
    }
    return result;
  }
}
