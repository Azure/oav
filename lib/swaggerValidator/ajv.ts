import { Ajv, CompilationContext } from "ajv";
import { JsonLoader } from "../swagger/jsonLoader";
import { Schema } from "../swagger/swaggerTypes";
import { xmsAzureResource, xmsMutability, xmsSecret } from "../util/constants";
import { ajvEnableDiscriminatorMap } from "./ajvDiscriminatorMap";

export const ajvEnableReadOnlyAndXmsMutability = (ajv: Ajv) => {
  ajv.removeKeyword("readOnly");
  ajv.addKeyword("readOnly", {
    metaSchema: { type: "boolean" },
    inline: (
      it: CompilationContext,
      _keyword: string,
      isReadOnly: boolean,
      parentSchema: Schema
    ) => {
      if (parentSchema?.[xmsMutability] !== undefined) {
        return "1";
      }
      const data = `data${it.dataLevel || ""}`;
      return isReadOnly ? `this.isResponse || ${data} === null || ${data} === undefined` : "1";
    },
  });

  ajv.addKeyword(xmsMutability, {
    metaSchema: { type: "array", items: { enum: ["create", "update", "read"] } } as Schema,
    inline: (
      it: CompilationContext,
      _keyword: string,
      mutability: Exclude<Schema[typeof xmsMutability], undefined>
    ) => {
      const validInRequest = mutability.includes("create") || mutability.includes("update");
      const validInResponse = mutability.includes("read");
      if (validInRequest && validInResponse) {
        return "1";
      }
      if (!validInRequest && !validInResponse) {
        throw new Error(`Invalid ${xmsMutability} value: ${JSON.stringify(mutability)}`);
      }
      const data = `data${it.dataLevel || ""}`;
      return `${
        validInRequest ? "!" : ""
      }this.isResponse || ${data} === null || ${data} === undefined`;
    },
  });
};

export const ajvEnableXmsSecret = (ajv: Ajv) => {
  ajv.addKeyword(xmsSecret, {
    metaSchema: { type: "boolean" } as Schema,
    inline: (it: CompilationContext, _keyword: string, isSecret: boolean) => {
      const data = `data${it.dataLevel || ""}`;
      return isSecret ? `!this.isResponse || ${data} === null || ${data} === undefined` : "1";
    },
  });
};

export const ajvEnableXmsAzureResource = (ajv: Ajv) => {
  ajv.addKeyword(xmsAzureResource, {
    metaSchema: { type: "boolean" } as Schema,
    inline: (it: CompilationContext, _keyword: string, isResource: boolean) => {
      const data = `data${it.dataLevel || ""}`;
      return isResource
        ? `!(this.isResponse && (this.httpMethod === 'get' || this.httpMethod === 'put')) || (${data}.id !== null && ${data}.id !== undefined)`
        : "1";
    },
  });
};

export const ajvEnableInt32AndInt64Format = (ajv: Ajv) => {
  ajv.addFormat("int32", {
    type: "number",
    validate: (x) => x % 1 === 0 && x >= -2_147_483_648 && x <= 2_147_483_647,
  });

  // TODO int64 range exceed Number.MAX_SAFE_INTEGER so we will lost precision when JSON.parse
  const int64Max = BigInt(2) ** BigInt(63) - BigInt(1);
  const int64Min = BigInt(2) ** BigInt(63) * BigInt(-1);
  ajv.addFormat("int64", {
    type: "number",
    validate: (x) => x % 1 === 0 && x >= int64Min && x <= int64Max,
  });
};

export const ajvEnableUnixTimeFormat = (ajv: Ajv) => {
  ajv.addFormat("unixtime", {
    type: "number",
    validate: (x) => x % 1 === 0,
  });
};

export const ajvAddFormatsDefaultValidation = (
  ajv: Ajv,
  type: "string" | "number",
  formats: string[]
) => {
  for (const format of formats) {
    ajv.addFormat(format, {
      type,
      validate: () => true,
    });
  }
};

export const ajvEnableDateTimeRfc1123Format = (ajv: Ajv) => {
  // https://tools.ietf.org/html/rfc822#section-5
  ajv.addFormat("date-time-rfc1123", {
    type: "string",
    validate:
      /^(?:(?:Mon|Tue|Wed|Thu|Fri|Sat|Sun), )?[0-3]\d (?:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec) \d\d(?:\d\d)? (?:[0-2]\d:[0-5]\d(?::[0-5]\d)|23:59:60) (?:[A-Z]{1,3})?(?:[+-]\d\d\d\d)?$/,
  });
};

export const ajvEnableDurationFormat = (ajv: Ajv) => {
  // https://en.wikipedia.org/wiki/ISO_8601#Durations
  ajv.addFormat("duration", {
    type: "string",
    validate:
      /^P([0-9]+(?:[,\.][0-9]+)?Y)?([0-9]+(?:[,\.][0-9]+)?M)?([0-9]+(?:[,\.][0-9]+)?D)?(?:T([0-9]+(?:[,\.][0-9]+)?H)?([0-9]+(?:[,\.][0-9]+)?M)?([0-9]+(?:[,\.][0-9]+)?S)?)?$/,
  });
};

// for (const keyword of [
//   "name",
//   "in",
//   "example",
//   "parameters",
//   "externalDocs",
//   "x-nullable",
//   "x-ms-enum",
//   "x-ms-azure-resource",
//   "x-ms-parameter-location",
//   "x-ms-client-name",
//   "x-ms-external",
//   "x-ms-skip-url-encoding",
//   "x-ms-client-flatten",
//   "x-ms-api-version",
//   "x-ms-parameter-grouping",
//   "x-ms-discriminator-value",
//   "x-ms-client-request-id",
//   "x-apim-code-nillable",
//   "x-new-pattern",
//   "x-previous-pattern",
//   "x-comment",
//   "x-abstract",
//   "allowEmptyValue",
//   "collectionFormat",
// ]) {
//   ajv.addKeyword(keyword, {});
// }

export const ajvEnableAll = (ajv: Ajv, jsonLoader: JsonLoader) => {
  ajvEnableDiscriminatorMap(ajv, jsonLoader);
  ajvEnableXmsSecret(ajv);
  ajvEnableReadOnlyAndXmsMutability(ajv);
  ajvEnableUnixTimeFormat(ajv);
  ajvEnableInt32AndInt64Format(ajv);
  ajvEnableDurationFormat(ajv);
  ajvEnableDateTimeRfc1123Format(ajv);
  ajvAddFormatsDefaultValidation(ajv, "string", ["byte", "password", "file", "base64url"]);
  ajvAddFormatsDefaultValidation(ajv, "number", ["double", "float", "decimal"]);
};

export const ajvEnableArmRule = (ajv: Ajv) => {
  ajvEnableXmsAzureResource(ajv);
};
