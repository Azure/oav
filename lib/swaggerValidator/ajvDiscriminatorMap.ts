import { default as Ajv, ErrorObject, ValidateFunction } from "ajv";
import { JsonLoader } from "../swagger/jsonLoader";
import { Schema } from "../swagger/swaggerTypes";

export const ajvEnableDiscriminatorMap = (ajv: Ajv, loader: JsonLoader) => {
  ajv.addKeyword({
    keyword: "discriminatorMap",
    errors: "full",
    metaSchema: {
      type: "object",
      additionalProperties: { type: "object", nullable: true },
    },

    compile(schemas: { [key: string]: Schema }, parentSch) {
      const parentSchema = parentSch as Schema;
      const compiled: { [key: string]: ValidateFunction | null } = {};
      const schemaMap: { [key: string]: Schema } = {};
      for (const value of Object.keys(schemas)) {
        compiled[value] = schemas[value] === null ? null : ajv.compile(schemas[value]);
        schemaMap[value] =
          schemas[value] === null ? parentSchema : loader.resolveRefObj(schemas[value]);
      }
      const discriminator = parentSchema.discriminator;
      if (discriminator === undefined) {
        throw new Error("Discriminator is absent");
      }
      const validated = new WeakSet<Schema>();
      const allowedValues = Object.keys(schemas);

      return function v(this: any, data: any, dataCxt) {
        let dataPath = dataCxt!.dataPath;
        if (data === null || data === undefined || typeof data !== "object") {
          // Should be validated by other schema property.
          return true;
        }
        dataPath = dataPath ?? "";
        const discriminatorValue = data[discriminator];
        const validate =
          discriminatorValue !== undefined && discriminatorValue !== null
            ? compiled[discriminatorValue]
            : undefined;

        if (validated.has(data)) {
          return true;
        }
        validated.add(data);
        const errors: ErrorObject[] = [];

        const sch = schemaMap[discriminatorValue] ?? parentSchema;

        if (validate === undefined || validate === null) {
          (v as any).errors = [
            {
              keyword: "discriminatorMap",
              data,
              dataPath,
              params: { allowedValues, discriminatorValue },
              schemaPath: "/discriminator",
              schema: schemas,
              parentSchema,
            } as ErrorObject,
          ];
          return false;
        }
        if (validate !== undefined && validate !== null) {
          const valid = validate.call(this, data);
          if (!valid && validate.errors) {
            for (const err of validate.errors) {
              err.dataPath = dataPath + err.dataPath;
            }
            errors.push(...validate.errors);
          }
        }

        if (
          sch.properties !== undefined &&
          sch.additionalProperties === undefined &&
          parentSchema.properties !== undefined &&
          parentSchema.additionalProperties === undefined
        ) {
          // We validate { additionalProperties: false } manually if it's not set
          for (const key of Object.keys(data)) {
            if (!(key in sch.properties)) {
              errors.push({
                keyword: "additionalProperties",
                data,
                dataPath,
                params: { additionalProperty: key },
                schemaPath: "/additionalProperties",
                schema: false,
                parentSchema,
              });
            }
          }
        }

        if (errors.length > 0) {
          (v as any).errors = errors;
          return false;
        } else {
          (v as any).errors = null;
          return true;
        }
      };
    },
  });
};
