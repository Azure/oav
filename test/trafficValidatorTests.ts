import * as path from "path";
import { TrafficValidator } from "../lib/swaggerValidator/trafficValidator";

describe("TrafficValidator", () => {
  it("validate data-plane traffic", async () => {
    const validator = new TrafficValidator(
      path.resolve(
        process.cwd(),
        "./test/liveValidation/swaggers/specification/cosmos-db/data-plane/Microsoft.Tables"
      ),
      path.resolve(process.cwd(), "./test/liveValidation/payloads/dataplane/")
    );
    await validator.initialize();
    const result = {} as any;
    result.validationIssues = await validator.validate();
    result.operationCoverage = validator.operationCoverageResult;
    expect(result).toMatchSnapshot();
  });
});
