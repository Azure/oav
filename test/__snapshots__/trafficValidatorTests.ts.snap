// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrafficValidator validate data-plane traffic 1`] = `
Object {
  "operationCoverage": Array [
    Object {
      "apiVersion": "2019-02-02",
      "coverageRate": 0.14285714285714285,
      "coveredOperations": 2,
      "totalOperations": 14,
      "unCoveredOperations": 12,
      "unCoveredOperationsList": Array [
        Object {
          "key": "Service",
          "operationId": "Service_GetProperties",
        },
        Object {
          "key": "Service",
          "operationId": "Service_GetStatistics",
        },
        Object {
          "key": "Service",
          "operationId": "Service_SetProperties",
        },
        Object {
          "key": "Table",
          "operationId": "Table_Create",
        },
        Object {
          "key": "Table",
          "operationId": "Table_DeleteEntity",
        },
        Object {
          "key": "Table",
          "operationId": "Table_GetAccessPolicy",
        },
        Object {
          "key": "Table",
          "operationId": "Table_InsertEntity",
        },
        Object {
          "key": "Table",
          "operationId": "Table_MergeEntity",
        },
        Object {
          "key": "Table",
          "operationId": "Table_QueryEntities",
        },
        Object {
          "key": "Table",
          "operationId": "Table_QueryEntityWithPartitionAndRowKey",
        },
        Object {
          "key": "Table",
          "operationId": "Table_SetAccessPolicy",
        },
        Object {
          "key": "Table",
          "operationId": "Table_UpdateEntity",
        },
      ],
      "unCoveredOperationsListGen": Array [
        Object {
          "operationIdList": Array [
            Object {
              "key": "Service",
              "operationId": "Service_SetProperties",
            },
            Object {
              "key": "Service",
              "operationId": "Service_GetProperties",
            },
            Object {
              "key": "Service",
              "operationId": "Service_GetStatistics",
            },
          ],
        },
        Object {
          "operationIdList": Array [
            Object {
              "key": "Table",
              "operationId": "Table_Create",
            },
            Object {
              "key": "Table",
              "operationId": "Table_QueryEntities",
            },
            Object {
              "key": "Table",
              "operationId": "Table_QueryEntityWithPartitionAndRowKey",
            },
            Object {
              "key": "Table",
              "operationId": "Table_UpdateEntity",
            },
            Object {
              "key": "Table",
              "operationId": "Table_MergeEntity",
            },
            Object {
              "key": "Table",
              "operationId": "Table_DeleteEntity",
            },
            Object {
              "key": "Table",
              "operationId": "Table_InsertEntity",
            },
            Object {
              "key": "Table",
              "operationId": "Table_GetAccessPolicy",
            },
            Object {
              "key": "Table",
              "operationId": "Table_SetAccessPolicy",
            },
          ],
        },
      ],
      "validationFailOperations": 1,
    },
  ],
  "validationIssues": Array [
    Object {
      "errors": Array [],
      "operationInfo": Object {
        "apiVersion": "2019-02-02",
        "operationId": "Table_Delete",
        "position": undefined,
      },
      "payloadFilePathPosition": Object {
        "column": 18,
        "line": 18,
      },
      "runtimeExceptions": Array [],
    },
    Object {
      "errors": Array [
        Object {
          "code": "MISSING_REQUIRED_PARAMETER",
          "documentationUrl": "",
          "issueSource": "request",
          "jsonPathsInPayload": Array [
            ".headers.dataserviceversion",
          ],
          "message": "Value is required but was not provided",
          "params": Array [
            "dataserviceversion",
          ],
          "pathsInPayload": Array [
            "/headers/dataserviceversion",
          ],
          "schemaPath": "#/properties/headers/required",
          "severity": 0,
          "source": Object {
            "jsonRef": "#/parameters/DataServiceVersionParameter",
            "position": Object {
              "column": 36,
              "line": 1762,
            },
            "url": "preview/2019-02-02/table.json",
          },
        },
        Object {
          "code": "INVALID_RESPONSE_CODE",
          "documentationUrl": "",
          "issueSource": "response",
          "jsonPathsInPayload": Array [],
          "message": "The swagger file does not define '204' response code",
          "pathsInPayload": Array [],
          "schemaPath": "",
          "severity": 0,
          "source": Object {
            "position": Object {
              "column": 22,
              "line": 80,
            },
            "url": "preview/2019-02-02/table.json",
          },
        },
      ],
      "operationInfo": Object {
        "apiVersion": "2019-02-02",
        "operationId": "Table_Query",
        "position": Object {
          "column": 24,
          "line": 43,
        },
      },
      "payloadFilePathPosition": Object {
        "column": 17,
        "line": 2,
      },
      "runtimeExceptions": Array [],
    },
  ],
}
`;
